R/use_extendr.R

```diff
@@ -181,4 +178,20 @@ make_example_wrappers <- function(pkg_name, outfile, extra_items = NULL) {
   }
 
   brio::write_lines(wrappers_content, outfile)
+  if (!isTRUE(quiet)) {
+    rel_path <- pretty_rel_path(outfile, search_from = path)
+    cli::cli_alert_success("Writting wrappers to {.file {rel_path}}.")
+  }
+}
+
+write_namespace <- function(pkg_name, path = ".", quiet = FALSE) {
```

Comment:
I would prefer if we didn't have this function at all. It feels like mission creep to me.

Reply:
Yeah, I agree. I did not actually check if we still need namespace modification. Before there were problems during the first `devtools::document()` -- during first package loading, `NAMESPACE` was missing `useDynLib`, and the native library was not linked, breaking all `.Call`s into it, including wrapper generation. With manually writing `useDynLib` into the `NAMESPACE`, first package loading also loads freshly compiled library, and all `.Call`s are OK. 

I'll let you know if we still need this.

Reply:
I am afraid we need to make sure `NAMESPACE` contains `useDynLib`, otherwise initial compilation will fail.
Here is a reprex whcih does not modify `NAMESPACE`:
<details>
<summary> Reprex </summary>

``` r
temp_dir <- tempdir()
pkg_dir <- file.path(temp_dir, "testpkg")
dir.create(pkg_dir, recursive = TRUE)
setwd(pkg_dir)

devtools::create(".")
#> v Setting active project to 'C:/Users/[redacted]/AppData/Local/Temp/RtmpeKV5JY/testpkg'
#> v Creating 'R/'
#> v Writing 'DESCRIPTION'
#> Package: testpkg
#> Title: What the Package Does (One Line, Title Case)
#> Version: 0.0.0.9000
#> Authors@R (parsed):
#>     * First Last <first.last@example.com> [aut, cre] (YOUR-ORCID-ID)
#> Description: What the package does (one paragraph).
#> License: `use_mit_license()`, `use_gpl3_license()` or friends to pick a
#>     license
#> Encoding: UTF-8
#> LazyData: true
#> Roxygen: list(markdown = TRUE)
#> RoxygenNote: 7.1.1
#> v Writing 'NAMESPACE'
#> v Setting active project to '<no active project>'
rextendr::use_extendr()
#> v Creating src/rust/src.
#> v Writing 'C:/Users/[redacted]/AppData/Local/Temp/RtmpeKV5JY/testpkg/src/entrypoint.c'
#> v Setting active project to '<no active project>'
#> v Writing 'C:/Users/[redacted]/AppData/Local/Temp/RtmpeKV5JY/testpkg/src/Makevars'
#> v Setting active project to '<no active project>'
#> v Writing 'C:/Users/[redacted]/AppData/Local/Temp/RtmpeKV5JY/testpkg/src/Makevars.win'
#> v Setting active project to '<no active project>'
#> v Writing 'C:/Users/[redacted]/AppData/Local/Temp/RtmpeKV5JY/testpkg/src/.gitignore'
#> v Setting active project to '<no active project>'
#> v Writing 'C:/Users/[redacted]/AppData/Local/Temp/RtmpeKV5JY/testpkg/src/rust/Cargo.toml'
#> v Setting active project to '<no active project>'
#> v Writing 'C:/Users/[redacted]/AppData/Local/Temp/RtmpeKV5JY/testpkg/src/rust/src/lib.rs'
#> v Setting active project to '<no active project>'
#> v Writting wrappers to R/extendr-wrappers.R.
#> v Finished configuring extendr for package testpkg.
#> ! Please update the system requirement in DESCRIPTION file.
#> ! Please run `rextendr::document()` for changes to take effect.
rextendr::document()
#> ! No library found at src/testpkg.dll, recompilation is required.
#> i Generating extendr wrapper functions for package: testpkg.
#> i Loading testpkg
#> Re-compiling testpkg
#>   -  installing *source* package 'testpkg' ...
#>      ** using staged installation
#>      ** libs
#>      rm -Rf testpkg.dll ./rust/target/x86_64-pc-windows-gnu/release/libtestpkg.a entrypoint.o
#>      "C:/rtools40/mingw64/bin/"gcc  -I"C:/PROGRA~1/R/R-40~1.4/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mfpmath=sse -msse2 -mstackrealign -UNDEBUG -Wall -pedantic -g -O0 -c entrypoint.c -o entrypoint.o
#>      cargo build --target=x86_64-pc-windows-gnu --lib --release --manifest-path=./rust/Cargo.toml
#>              Updating crates.io index
#>             Compiling winapi-build v0.1.1
#>       Compiling winapi-x86_64-pc-windows-gnu v0.4.0
#>       Compiling winapi v0.3.9
#>       Compiling winapi v0.2.8
#>       Compiling proc-macro2 v1.0.24
#>       Compiling unicode-xid v0.2.1
#>       Compiling syn v1.0.63
#>             Compiling extendr-engine v0.2.0
#>             Compiling lazy_static v1.4.0
#>             Compiling kernel32-sys v0.2.2
#>             Compiling quote v1.0.9
#>             Compiling extendr-macros v0.2.0
#>             Compiling libR-sys v0.2.1
#>             Compiling extendr-api v0.2.0
#>             Compiling testpkg v0.1.0 (C:\Users\i[redacted]\AppData\Local\Temp\RtmpeKV5JY\testpkg\src\rust)
#>              Finished release [optimized] target(s) in 36.67s
#>      C:/rtools40/mingw64/bin/gcc -shared -s -static-libgcc -o testpkg.dll tmp.def entrypoint.o -L./rust/target/x86_64-pc-windows-gnu/release -ltestpkg -lws2_32 -ladvapi32 -luserenv -LC:/PROGRA~1/R/R-40~1.4/bin/x64 -lR
#>      installing to C:/Users/[redacted]/AppData/Local/Temp/Rtmpy8Equp/devtools_install_5df82b6e6ce7/00LOCK-testpkg/00new/testpkg/libs/x64
#>   -  DONE (testpkg)
#> Error in .Call(wrapper_function, use_symbols = use_symbols, package_name = package_name,  : 
#>   "wrap__make_testpkg_wrappers" not available for .Call() for package "testpkg"
#> Error: Failed to generate wrapper functions
```

<sup>Created on 2021-03-13 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

</details>

Reply:
I feel this is a problem in the function that is making the wrappers. It should silently (or with a warning) ignore the fact that we don't have a wrapper generating function, instead of erroring out.

Reply:
I may be wrong here, but without `NAMESPACE` entry we will not be able to compile code and generate wrappers in one call to `rextendr::document()`. Even if we somehow bypass this, by the end of the first compilation no exported function would work: even though the library is compiled, it is not dynamically loaded when the package is loaded (because of the missing `useDynLoad`). This is exactly what happens when we call `pkgload::load_all` and then attempt to get wrappers -- the library is compiled, the package is loaded, but without the library. No `.Call`s work. If we ensure that `NAMESPACE` contains `useDynLib`, we will be able to make `.Call`s and generate wrappers. This is what currently happens.

Basically, we write to `NAMESPACE` before it gets updated by {roxygen2}. After the first `rextendr::document()` the `NAMESPACE` gets rewritten and we get a `useDynLib` entry because we have an attribute in wrappers. After this, we never ever touch `NAMESPACE` again.

I have no better solution and am open to suggestions.

Reply:
I was thinking the following, starting from an empty project

1. `rextendr::use_extendr()`: Basic wrapper file gets written
2. `rextendr::document()` (or alternatively, `devtools::document()`, it's the same here): `NAMESPACE` gets updated, this transfers `useDynLoad` into `NAMESPACE`
3. `devtools::install()`: dynamic library gets built for the first time

Reply:
FWIW, with cpp11, it's a user's job to ensure `@useDynLib` is in `NAMAESPACE` (via the package doc). So, I'm wondering if we are doing too much.

```
> usethis::use_cpp11()
✓ Creating 'src/'
✓ Adding '*.o', '*.so', '*.dll' to 'src/.gitignore'
● Copy and paste the following lines into 'R/<package name>-package.R':
  ## usethis namespace: start
  #' @useDynLib <package name>, .registration = TRUE
  ## usethis namespace: end
  NULL
  [Copied to clipboard]
✓ Adding 'cpp11' to LinkingTo field in DESCRIPTION
✓ Adding 'C++11' to SystemRequirements field in DESCRIPTION
✓ Writing 'src/code.cpp'
● Modify 'src/code.cpp'
```

To do this programmatically, we need to tweak `NAMESPACE` anyway, though.

https://cpp11.r-lib.org/reference/cpp_register.html#examples

Reply:
Yes, I'm aware, and originally I didn't like that extendr writes `@useDynLib` into the wrappers, but now I think it's probably the right approach. I'm not convinced that the cpp way is done for some fundamental reasons based in first principles. It's most likely historic, that Rcpp was written before people even were widely using roxygen2. If you started from scratch, and you had some autogenerated code, wouldn't you want that code to include all the boilerplate that is needed?

In the end, if it comes down to depending on usethis for a convenient user experience, I'm fine with going that route. I just want to make sure we are somewhat careful in the dependencies we add. But any dependency that adds real value is fine by me.

Reply:
@clauswilke, @yutannihilation,
I think I am not entirely following the discission.
1. `rextendr::use_extendr()` creates wrappers (currently it writes `@useDynLib`)
2. `devtools::document()` compiles code, updates `NAMESPACE`, loads package.
3. But, to get **correct** `NAMESPACE`, we need to compile, then load package and `.Call` into wrapper generation function, which at this moment does not work -- `NAMESPACE` has not been updated and `useDynLib()` is missing. The alternative is not using `pkgload::load_all` but something like `pkgbuild::build()` and then manually loading the library with `dyn.load`, creating up to date wrappers, unloading, and then calling `devtools::document`, which will consume correct wrappers with all the necessary {roxygen2} tags. That way we do not need to write `NAMESPACE`.

If you have a better solution or at least a tiny reprex of how `rextendr:document()` should work, please share or make a PR in this branch.

Reply:
Can we dyn.load() the lib file directly?

Reply:
@yutannihilation, yes. We basically do this in {rextendr} in `rust_source()` and similar functions, which generate libraries on the fly from user input ({Rcpp}/{cpp1} should do the same). The challenge is to reliably load and **unload**, otherwise, the dll will get stuck (at least on Windows).
There is `devtools::unload()` that I think unloads everything from everywhere. Alternatively, the fact we use {callr} means that by the end of the `callr::r()` another process has already finished, and the library is naturally unloaded.

Reply:
@Ilia-Kosenkov 
Here's my take:

https://github.com/Ilia-Kosenkov/rextendr/pull/2

I believe this is the same idea as what you suggested here (sorry, I typed the comment above before viewing this).

> 3\. The alternative is not using `pkgload::load_all` but something like `pkgbuild::build()` and then manually loading the library with `dyn.load`, creating up to date wrappers, unloading, and then calling `devtools::document`, which will consume correct wrappers with all the necessary {roxygen2} tags. That way we do not need to write `NAMESPACE`.

Reply:
@yutannihilation,
Currently `pkgbuild::compile_dll` does not re-compile if `pkgbuild::needs_compile()` returns false.
I suspect there are other path related issues as well, which we do not uncover with the CI tests.
I am investigating this right now.

Reply:
Ah, I didn't notice it. Do you mean I should have added `force = TRUE`?

Reply:
@yutannihilation, Yeah, check the last two commits here. If you have time, feel free to checkout this PR and try out new package creation & building.